script/github-actions/check-deployability.js

```diff
@@ -0,0 +1,174 @@
+ */
+const isAncestor = (commitA, commitB) => {
+  const exitCode = runCommandSync(
+    `git merge-base --is-ancestor ${commitA} ${commitB}`,
+  );
+
+  if (![0, 1].includes(exitCode)) {
+    throw new Error(
+      `'git merge-base --is-ancestor' exited with an unsuccessful code (${exitCode}).`,
+    );
+  }
+
+  return commitA !== commitB && exitCode === 0;
+};
+
+/**
+ * Checks if the GITHUB_SHA is ahead of the last full deploy of an environment.
+ *
+ * @param {string} env - Environment name
+ * @returns {Boolean} Returns true if GITHUB_SHA is ahead of the last full deploy.
+ */
+const isAheadOfLastFullDeploy = async env => {
+  const lastFullDeployCommit = await getLastFullDeployCommit(env);
+  return isAncestor(GITHUB_SHA, lastFullDeployCommit);
+};
+
+/**
+ * Determines if the GITHUB_SHA can be deployed to non production environments.
+ *
+ * @param {string} env - Environment name
+ * @returns {Boolean} Whether or not the GITHUB_SHA can be deployed to the environment.
+ */
+const checkDeployability = async env => {
+  if (!(await isAheadOfLastFullDeploy(env))) return false;
+
+  const inProgressWorkflowRuns = await getInProgressWorkflowRuns(
+    'continuous-integration.yml',
+  );
+  if (inProgressWorkflowRuns.length === 1) return true;
+
+  const previousCommitsInProgress = inProgressWorkflowRuns.find(workflowRun =>
+    isAncestor(workflowRun.head_sha, GITHUB_SHA),
+  );
+
+  if (!previousCommitsInProgress) return true;
+
+  const timeout = 5; // Number of minutes to wait before checking again
+  console.log('Waiting for previous workflow runs to finish deploying...');
+  await sleep(timeout * 60 * 1000);
+
+  return checkDeployability(env);
+};
+
+/**
+ * Determines whether the GITHUB_SHA can be deployed to production. This is
+ * intended to be used for isolated app commits in the `main` branch
+ * to avoid a race condition with the daily production deploy.
+ *
+ * @returns {Boolean} Whether or not the GITHUB_SHA can be deployed to production.
+ */
+const checkDeployabilityProd = async () => {
+  if (!(await isAheadOfLastFullDeploy(ENVIRONMENTS.VAGOVPROD))) return false;
+
+  const inProgressWorkflowRuns = await getInProgressWorkflowRuns(
+    'daily-deploy-production.yml',
+  );
+  if (inProgressWorkflowRuns.length === 0) return true;
+
+  // Get the first item in the Array. Since workflow runs for the daily production
+  // deploy aren't concurrent, there should be only one deploy happening at a time.
+  const dailyDeploySha = inProgressWorkflowRuns[0].head_sha;
+
+  // Don't deploy isolated app commits that are older than the daily deploy
+  // commit. The daily deploy will include the changes from the older commit.
+  const isAheadOfDailyDeploy = isAncestor(GITHUB_SHA, dailyDeploySha);
+  if (!isAheadOfDailyDeploy) return false;
+
+  const timeout = 10; // Number of minutes to wait before checking again
+  console.log('Waiting for the Daily Production Deploy to complete...');
```

Comment:
I like this log output, and you could consider adding more logging in other areas along the lines of "Daily production deploy already running with newer commit, skipping single-app deploy". That way it's clear to anyone watching the GHA output exactly what happened (or didn't happen) and why

Reply:
That's a good point, especially since teams may be wondering why their commit skipped deployment. I'll also add a log to the `isDeployableToEnv` function for when it skips the deploys for older commits